hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,195 @@
+	typeWalker.originalVisitObjectType = visitor.VisitObjectType
+
+	// visitor is a copy - modifications won't impact passed visitor
+	visitor.VisitTypeName = typeWalker.visitTypeName
+	visitor.VisitObjectType = typeWalker.visitObjectType
+
+	typeWalker.visitor = visitor
+	typeWalker.AfterVisitFunc = DefaultAfterVisit
+	typeWalker.MakeContextFunc = IdentityMakeContext
+	typeWalker.WalkCycle = IdentityWalkCycle
+
+	return &typeWalker
+}
+
+func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+	updatedCtx, err := t.MakeContextFunc(it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+	var ok bool
+	it, ok = visitedTypeName.(TypeName)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+	}
+
+	def, ok := t.allTypes[it]
+	if !ok {
+		return nil, errors.Errorf("couldn't find type %q", it)
+	}
+
+	// Prevent loops by bypassing this type if it's currently being processed. The processing
+	// slice is basically the "path" taken to get to the current type.
+	if _, ok := t.state.processing[def.Name()]; ok {
+		return t.WalkCycle(def, updatedCtx)
+	}
+	t.state.processing[def.Name()] = struct{}{}
+
+	updatedType, err := this.Visit(def.Type(), updatedCtx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error visiting type %q", def.Name())
+	}
+	updatedDef, err := t.AfterVisitFunc(def, def.WithType(updatedType), updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	delete(t.state.processing, def.Name())
+
+	err = t.state.result.AddWithEqualityCheck(updatedDef)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedDef.Name(), nil
+}
+
+func (t *TypeWalker) visitObjectType(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error) {
+	result, err := t.originalVisitObjectType(this, it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ot, ok := result.(*ObjectType)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitObjectType must return a *ObjectType, instead returned %T", result))
+	}
+
+	for _, prop := range ot.Properties() {
+		isCycle := isRemoveType(prop.PropertyType())
+		if isCycle {
+			ot = ot.WithoutProperty(prop.PropertyName())
+		}
+	}
+
+	return ot, nil
+}
+
+// Walk returns a Types collection constructed by applying the Visitor to each type in the graph of types reachable
+// from the provided TypeDefinition 'def'. Types are visited in a depth-first order. Cycles are not visited.
+func (t *TypeWalker) Walk(def TypeDefinition, ctx interface{}) (Types, error) {
+	t.state = typeWalkerState{
+		result:     make(Types),
+		processing: make(map[TypeName]struct{}),
+	}
+
+	t.state.processing[def.Name()] = struct{}{}
+
+	updatedType, err := t.visitor.Visit(def.Type(), ctx)
+	if err != nil {
+		return nil, err
+	}
+	updatedDef, err := t.AfterVisitFunc(def, def.WithType(updatedType), ctx)
```

Comment:
Should there be a call to `BeforeVisit` before the visit? Or is this why the callback shouldn't be called `BeforeVisit`?

Reply:
Hmm, yeah I see your point here.

Since `MakeContextFunc` is all about managing the `ctx` for each type in the graph, it's not called here because the seed `ctx` is used for the first call. The N+1st call uses a `ctx` created by calling `MakeContextFunc(nthCtx)`.

Reply:
As discussed, changed it so that there is no context on `Walk` and the context is always manufactured by `MakeContext`. This allowed me to reduce a bit of duplication between `Walk` and `visitTypeName`. It does mean that the `MakeContext` function needs to handle its own base case (if it's nil) though, which does push a small bit of complexity to the caller.

I think it makes "where did this context come from" easier to reason about though.